src/vsr/README.md

```diff
@@ -0,0 +1,184 @@
+
+This protocol approach enables liveness under asymmetric network partitions. For example, a replica which can send to the cluster but not receive may send SVCs, but if the remainder of the cluster is healthy, they will never achieve a quorum, so the view is stable. When the partition heals, the formerly-isolated replica may rejoin the original view (if it was isolated in `status=normal`) or a new view (if it was isolated in `status=view_change`).
+
+See also:
+
+  - [Raft does not Guarantee Liveness in the face of Network Faults](https://decentralizedthoughts.github.io/2020-12-12-raft-liveness-full-omission/) ("PreVote and CheckQuorum")
+  - ["Consensus: Bridging Theory and Practice"](https://web.stanford.edu/~ouster/cgi-bin/papers/OngaroPhD.pdf) §6.2 "Leaders" describes periodically committing a heartbeat to detect stale leaders.
+
+## Protocol: View-Change
+
+A replica sends `command=do_view_change` to all replicas, with the `view` it is attempting to start.
+- The _primary_ of the `view` collects a [view-change quorum](#quorums) of DVCs.
+- The _backup_ of the `view` uses to `do_view_change` to updates its current `view` (transitioning to `status=view_change`).
+
+When the primary collects its quorum:
+1. The primary installs the headers to its suffix.
+2. Then the primary repairs (and potentially truncates uncommitted ops).
+3. When the primary finishes repair, it transitions to `status=normal` and broadcasts a `command=start_view`.
+
+## Protocol: Request/Start View
+
+### `request_start_view`
+
+A backup sends a `command=request_start_view` to the primary of a view when any of the following occur:
+
+  - the backup learns about a newer view via a `command=commit` message, or
+  - the backup learns about a newer view via a `command=prepare` message, or
+  - a replica recovers to `status=recovering_head`
+
+### `start_view`
+
+When a `status=normal` primary receives `command=request_start_view`, it replies with a `command=start_view`.
+`command=start_view` includes the view's current suffix — the headers of the latest messages in the view.
+
+Upon receiving a `start_view` for the new view, the backup installs the suffix, transitions to `status=normal`, and begins repair.
+
+## Protocol: Repair Journal
+
+`request_headers` and `headers` repair gaps or breaks in a replica's journal headers.
+Because the headers are repaired backwards (from the head) by hash-chaining, it is safe for both backups and transitioning primaries.
+
+## Protocol: Repair WAL
+
+In response to a `request_prepare`:
+
+- Reply the `command=prepare` with the requested prepare, if available.
+- Reply `command=nack_prepare` if the request origin is the primary of the ongoing view-change and we never received the prepare. (This enables the primary to truncate uncommitted messages and remain available).
```

Comment:
This is safe, because if we already are `status=view_change`, we won't accept the prepare even if it arrives later (and view_durable check on sending a message ensures that this logic holds even acrosss restarts). 

Might be worth mentioning this, but I didn't find a clean way to do that